test/testdata/resolver/final_module.rb

```diff
@@ -0,0 +1,58 @@
+# typed: true
+# disable-fast-path: true
+
+class FinalClass
+  extend T::Helpers
+  final!
+end
+
+module FinalModule
+  extend T::Helpers
+  final!
+end
+
+class BadInherit < FinalClass # error: `FinalClass` was declared as final and cannot be inherited by `BadInherit`
+end
+
+module BadInclude # error: `FinalModule` was declared as final and cannot be included by `BadInclude`
+  include FinalModule
+end
+
+module BadExtend # error: `FinalModule` was declared as final and cannot be extended by `BadExtend`
+  extend FinalModule
+end
```

Comment:
i don't think your PR fails for this currently, but can you add a test which mixes and matches includes / extends of final / non-final modules?